selfservice/strategy/oidc/strategy_login.go

```diff
@@ -167,22 +167,49 @@ func (s *Strategy) processLogin(w http.ResponseWriter, r *http.Request, loginFlo
 	var oidcCredentials identity.CredentialsOIDC
 	if err := json.NewDecoder(bytes.NewBuffer(c.Config)).Decode(&oidcCredentials); err != nil {
-		return nil, s.handleError(w, r, loginFlow, provider.Config().ID, nil, errors.WithStack(herodot.ErrInternalServerError.WithReason("The password credentials could not be decoded properly").WithDebug(err.Error())))
+		return nil, s.handleError(w, r, loginFlow, provider.Config().ID, nil, errors.WithStack(herodot.ErrInternalServerError.WithReason("The OpenID connect credentials could not be decoded properly").WithDebug(err.Error())))
 	}
 
 	sess := session.NewInactiveSession()
 	sess.CompletedLoginForWithProvider(s.ID(), identity.AuthenticatorAssuranceLevel1, provider.Config().ID,
 		httprouter.ParamsFromContext(r.Context()).ByName("organization"))
-	for _, c := range oidcCredentials.Providers {
-		if c.Subject == claims.Subject && c.Provider == provider.Config().ID {
-			if err = s.d.LoginHookExecutor().PostLoginHook(w, r, node.OpenIDConnectGroup, loginFlow, i, sess, provider.Config().ID); err != nil {
-				return nil, s.handleError(w, r, loginFlow, provider.Config().ID, nil, err)
-			}
-			return nil, nil
+
+	found := -1
+	for k := range oidcCredentials.Providers {
+		p := &oidcCredentials.Providers[k]
+		if p.Subject == claims.Subject && p.Provider == provider.Config().ID {
+			found = k
+			break
 		}
 	}
 
-	return nil, s.handleError(w, r, loginFlow, provider.Config().ID, nil, errors.WithStack(herodot.ErrInternalServerError.WithReason("Unable to find matching OpenID Connect Credentials.").WithDebugf(`Unable to find credentials that match the given provider "%s" and subject "%s".`, provider.Config().ID, claims.Subject)))
+	if found == -1 {
+		return nil, s.handleError(w, r, loginFlow, provider.Config().ID, nil, errors.WithStack(herodot.ErrInternalServerError.WithReason("Unable to find matching OpenID Connect Credentials.").WithDebugf(`Unable to find credentials that match the given provider "%s" and subject "%s".`, provider.Config().ID, claims.Subject)))
+	}
+
+	if err = s.d.LoginHookExecutor().PostLoginHook(w, r, node.OpenIDConnectGroup, loginFlow, i, sess, provider.Config().ID); err != nil {
+		return nil, s.handleError(w, r, loginFlow, provider.Config().ID, nil, err)
+	}
+
+	if err := s.d.IdentityManager().UpdateCredentials(r.Context(), i.ID, identity.CredentialsTypeOIDC, func(toUpdate *identity.Credentials) error {
+		if err := toUpdate.UnmarshalConfig(toUpdate); err != nil {
+			return err
+		}
+		var toUpdateConfig = oidcCredentials
+		k, found := toUpdateConfig.GetProvider(provider.Config().ID, claims.Subject)
+		if !found {
+			// Credentials are not found, we can ignore this.
+			return nil
+		}
+		toUpdateConfig.Providers[k].CurrentIDToken = token.GetIDToken()
+		toUpdateConfig.Providers[k].CurrentAccessToken = token.GetAccessToken()
+		toUpdateConfig.Providers[k].CurrentRefreshToken = token.GetRefreshToken()
+		toUpdate.Config, err = json.Marshal(toUpdateConfig)
+		return errors.WithStack(err)
```

Comment:
I think this code has a few problems:

1. `toUpdate.UnmarshalConfig(toUpdate)` unmarshalls on itself?
2. To update the identity we are fetching it again (within UpdateCredentials) which will mean + ~200ms for every login. We already have the correct credentials from "FindByCredentials" so let's use that credentials instead.
3. Since this adds an update to every login it can increase the login time on some databases by up to 500ms. It would be great if this was behind a feature flag and maybe even per provider (`selfservice.methods.oidc.providers.0.update_tokens_on_login`)
4. Please add ample tests (maybe move this into a separate function) that ensures that the credentials are correctly updated. If there is a bug, it could break login for every user using social sign in (not saying it has a bug, but tests make sure we don't have any!

Reply:
1. done
2. in progress
3. in progress
4. in progress

Reply:
That probably won't work well, because the request context will be canceled by then, which will lead to a rollback in the transaction. In my view it needs to be part of the request chain